clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1796,6 +1824,26 @@ void SYCLIntegrationHeader::emitForwardClassDecls(
         }
         break;
       }
+      case TemplateArgument::ArgKind::Integral: {
+        // Handle Kernel Name Type templated using enum.
+        QualType T = Arg.getIntegralType();
+        if (const EnumType *ET = T->getAs<EnumType>()) {
+          auto EnumDecl = ET->getDecl();
+          DeclContext *DC = EnumDecl->getDeclContext();
+          auto *NS = dyn_cast_or_null<NamespaceDecl>(DC);
+          bool InCLNameSpace = false;
+
+          // Header files are included before integration header. Enums
+          // therefore defined in headers need not be forward declared. Their
+          // definitions are available in integration header.
+          if (NS)
+            InCLNameSpace = checkIfDeclInCLNameSpace(DC);
```

Comment:
What if a user declared enum in cl namespace and used it in kernel name?
I guess it is not prohibited by the spec.

Just wondering, can we somehow check that definition of enum is available in the integration header instead of checking that it is in `cl` namespace.

Reply:
Is it NOT prohibited by the spec? The C++ Spec prohibits adding anything to the std namespace, it would be a terrible mistake if SYCL didn't do the same to its namespaces.

Reply:
I haven't seen statements in the SYCL spec which prohibit using of `cl` or `cl::sycl` namespace. The only one related statement I've seen is (pt. 4.1):
> The `cl::sycl::detail` namespace is reserved for implementation details.

I can miss something, tagging @rolandschulz @mkinsner .

Reply:
Oh good grief.  @rolandschulz  @mkinsner  THIS NEEDS TO BE FIXED ASAP!

The sentence above needs to be expanded to the cl and cl::sycl namespace.  Otherwise adding anything to those namespace becomes a breaking change.

Reply:
Filed https://github.com/KhronosGroup/SYCL-Docs/issues/97 . I would proceed assuming that this will be clarified that the user isn't allowed to put anything into cl or sycl namespace.